### PR TITLE
home-assistant: add lifx component dependencies

### PR DESCRIPTION
**Pull request by netixx**
###### Motivation for this change
This commit solves https://github.com/NixOS/nixpkgs/issues/55527 by adding required
python packages and binding them in the component-packages.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by netixx** [comment]
Thank you @dotlambda for helping to debug my setup and making this PR possible !

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ stdenv,
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
{ lib
```

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  fetchPypi,
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
, fetchPypi
```

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "cf53c9faea6eee25a466e73eef1753b82a75c7497648149c19c15342df2678f2";
+  };
+
+  // tests are not implemented :
+  // nix_run_setup: error: unrecognized arguments: test
+  doCheck = false;
+
+  nativeBuildInputs = [ ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/aiolifx_effects/default.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{ stdenv,
+  fetchPypi,
+  buildPythonPackage,
+  aiolifx
+}:
+
+buildPythonPackage rec {
+  pname = "aiolifx_effects";
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx_effects/default.nix]:
```suggestion
  pname = "aiolifx-effects";
```

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -168,6 +168,10 @@ in {
 
   aioimaplib = callPackage ../development/python-modules/aioimaplib { };
 
+  aiolifx = callPackage ../development/python-modules/aiolifx { };
+
+  aiolifx_effects = callPackage ../development/python-modules/aiolifx_effects { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]:
```suggestion
  aiolifx-effects = callPackage ../development/python-modules/aiolifx-effects { };
```
See https://nixos.org/nixpkgs/manual/#contributing.

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  pname = "aiolifx";
+  version = "0.6.7";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "cf53c9faea6eee25a466e73eef1753b82a75c7497648149c19c15342df2678f2";
+  };
+
+  // tests are not implemented :
+  // nix_run_setup: error: unrecognized arguments: test
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+buildPythonPackage rec {
+  pname = "aiolifx";
+  version = "0.6.7";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "cf53c9faea6eee25a466e73eef1753b82a75c7497648149c19c15342df2678f2";
+  };
+
+  // tests are not implemented :
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
  # tests are not implemented
```

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  // nix_run_setup: error: unrecognized arguments: test
+  doCheck = false;
+
+  nativeBuildInputs = [ ];
+  propagatedBuildInputs = [ bitstring ifaddr ];
+
+  meta = with stdenv.lib; {
+    homepage = "http://github.com/frawau/aiolifx";
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]:
```suggestion
    homepage = http://github.com/frawau/aiolifx;
```

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    license = licenses.mit;
+    description = "API for local communication with LIFX devices over a LAN with asyncio.";
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]: https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please make a separate commit for each package and add a `pythonPackages.` prefix to the commit messages.

**Comment by dotlambda** [comment]
From https://github.com/frawau/aiolifx:
> aiolifx is a **Python 3**/asyncio library to control Lifx LED lightbulbs over your LAN.

File: pkgs/development/python-modules/aiolifx_effects/default.nix (hunk 4 above)
Comment (netixx) [reply pkgs/development/python-modules/aiolifx_effects/default.nix]: I originally did that, but I would get a 404 error when using a `-`. Did you have success with this syntax ?

File: pkgs/development/python-modules/aiolifx/default.nix (hunk 9 above)
Comment (netixx) [reply pkgs/development/python-modules/aiolifx/default.nix]: should I add myself to the maintainers in a separate (first) commit ?

**Comment by netixx** [comment]
> From https://github.com/frawau/aiolifx:
> 
> > aiolifx is a **Python 3**/asyncio library to control Lifx LED lightbulbs over your LAN.

How do I transcribe this in the nix package manager ?

**Comment by netixx** [comment]
Also, I included only `aiolifx_effects` (or `aiolifx-effects`) in the component dependencies, because `aiolifx` is a dependency, but the `lifx` components also uses the `aiolifx` directly (i.e. there are `import aiolifx` statements) so in a way it also has a direct dependency.

Should I add `aiolifx` in the component dependencies as well ?

File: pkgs/development/python-modules/aiolifx_effects/default.nix (hunk 4 above)
Comment (dotlambda) [reply pkgs/development/python-modules/aiolifx_effects/default.nix]: You have to add `pname = "aiolifx_effects";` inside `fetchPypi`.

File: pkgs/development/python-modules/aiolifx/default.nix (hunk 9 above)
Comment (dotlambda) [reply pkgs/development/python-modules/aiolifx/default.nix]: Yes, please.

**Comment by dotlambda** [comment]
> > From https://github.com/frawau/aiolifx:
> > > aiolifx is a **Python 3**/asyncio library to control Lifx LED lightbulbs over your LAN.
> 
> How do I transcribe this in the nix package manager ?

`disabled = !isPy3k`

> Also, I included only `aiolifx_effects` (or `aiolifx-effects`) in the component dependencies, because `aiolifx` is a dependency, but the `lifx` components also uses the `aiolifx` directly (i.e. there are `import aiolifx` statements) so in a way it also has a direct dependency.
> 
> Should I add `aiolifx` in the component dependencies as well ?

Please read the first two lines of `component-packages.nix`. You need to run `parse-requirements.py`.

**Comment by netixx** [comment]
I fixed what you suggested.

I have a chicken and egg problem with running `parse-requirements.py` because I need the setup with the pythonPackages commit (and my current setup already runs on a custom branch).

Do you have a magic onliner (or few liners) that I can use to quickly generate the component-packages.nix ?

**Comment by dotlambda** [comment]
I don't get what your problem is. Once you fixed all evaluation errors, there should be no problem running the script from your `hass-lifx-fix` branch.

File: pkgs/top-level/python-packages.nix [hunk 10]
```diff
@@ -299,7 +303,7 @@ in {
   cdecimal = callPackage ../development/python-modules/cdecimal { };
 
   cfn-flip = callPackage ../development/python-modules/cfn-flip { };
- 
+
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: please don't change random lines

File: pkgs/development/python-modules/aiolifx-effects/default.nix [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  src = fetchPypi {
+    inherit version;
+    pname = "aiolifx_effects";
+    sha256 = "cb4ac52deeb220783fc6449251cf40833fcffa28648270be64b1b3e83e06b503";
+  };
+
+  # tests are not implemented
+  doCheck = false;
+
+  disabled = !isPy3k
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx-effects/default.nix]: ;

File: pkgs/development/python-modules/aiolifx/default.nix [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "cf53c9faea6eee25a466e73eef1753b82a75c7497648149c19c15342df2678f2";
+  };
+
+  # tests are not implemented
+  doCheck = false;
+
+  disabled = !isPy3k
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx/default.nix]: ;

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 10 above)
Comment (netixx) [reply pkgs/top-level/python-packages.nix]: sorry my editor did this automatically when I saved the file (whitespace).

File: pkgs/development/python-modules/aiolifx-effects/default.nix (hunk 11 above)
Comment (netixx) [reply pkgs/development/python-modules/aiolifx-effects/default.nix]: sorry for being such a drag... I missed this during all my rebasing...

**Comment by netixx** [comment]
I wasn't sure it was going to work on my mac, but it did! However, I ran the script and another dependency was updated (`sensor.shodan`). I committed both for consistency but I will remove it if you think it's more appropriate.

File: pkgs/development/python-modules/aiolifx-effects/default.nix [hunk 13]
```diff
@@ -0,0 +1,31 @@
+{ lib
+  ,fetchPypi
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx-effects/default.nix]:
```suggestion
, fetchPypi
```

File: pkgs/development/python-modules/aiolifx-effects/default.nix [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  doCheck = false;
+
+  disabled = !isPy3k;
+
+  propagatedBuildInputs = [ aiolifx ];
+
+  meta = with lib; {
+    homepage = https://github.com/amelchio/aiolifx_effects;
+    license = licenses.mit;
+    description = "aiolifx light effects";
```
Comment (dotlambda) [inline pkgs/development/python-modules/aiolifx-effects/default.nix]: Please be more descriptive and follow https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
> I committed both for consistency

That's fine with me.

File: pkgs/development/python-modules/aiolifx-effects/default.nix (hunk 14 above)
Comment (netixx) [reply pkgs/development/python-modules/aiolifx-effects/default.nix]: I used the description on the github to provide more compete explanation.

File: pkgs/development/python-modules/aiolifx-effects/default.nix (hunk 13 above)
Comment (netixx) [reply pkgs/development/python-modules/aiolifx-effects/default.nix]: done (for all other as well).

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python3.pkgs.aiolifx-effects

**Comment by netixx** [comment]
Thank you for your help doing this !

**Pull request merged**